docs/local_assembly.tex

```diff
@@ -58,12 +58,19 @@ \section{Correcting Reads} \label{correcting-reads}
+The first step is to create a set of sequences to be kmerized and put into the de Bruijn graph.  These sequences are: the reference haplotype, any requested alleles given by the \code{alleles} argument in \HC's ``genotype given alleles" (GGA) mode\footnote{This mode is disabled in \Mutect~ and incompatible with the reference confidence mode of \HC, which is the recommended best practice.  We document it here for completeness.  Before assembly, the GATK engine creates GGA haplotypes by substituting the reference bases with the alt bases from the GGA vcf in the reference haplotype.  These GGA haplotypes are treated as non-reference sequences, that is, as long reads.}, and maximal subsequences of reads with base quality at least 10 by default\footnote{This is controlled by the \code{minBaseQualityScore} argument.} and at least one kmer long.  That is, a read with 100 bases and a base of quality 7 at position 70 yields sequences of length 69, 30 if $k = 10$ and a single sequence of length 69 if $k = 40$.  We do not use mate information; that is, a read and its mate yield completely independent sequences, nor do we later post-process the graph using mate information.  This is a potential direction for improvement.
 
+Before building the graph, the \code{ReadThreadingAssembler} finds the set of all kmers that appear more than once in the same sequence i.e. more than once in the reference or in any read.  These are called ``non-unique'' kmers and a kmer that is not non-unique is called ``unique.''  Then each sequence is kmerized starting from the first unique kmer\footnote{Note that a kmer that occurs only once in this particular sequence but more than once in some other sequence in non-unique.}, with the exception that the reference sequence begins at its first kmer, unique or not.  When a new kmer \textit{or a non-unique kmer (even if it is already on the graph)} appears, a new vertex is added to the graph; otherwise an edge is added (or the multiplicity of an existing edge incremented) from the preceding kmer to the current one.  The advantage of creating a new vertex for a non-unique kmer rather than re-using an old vertex is that it avoids spurious cycles.  Up to this point, we have strayed from a canonical de Bruijn graph only as noted above.  Note that this is a directed graph of kmers from aligned reads (i.e. aligned to the forward strand of the reference), not a bidirected graph of unaligned reads that could come from either strand, with the complications of associating a kmer with its reverse complement.
```

Comment:
> When a new kmer appears, a new vertex is added to the graph; otherwise an edge is added...

The otherwise is only true for unique kmers as I guess otherwise means an old kmer appears; old non-unique kmers may add a vertex if an edge from the previous kmer vertex with that extending base does not exists (adds a vertex and an edge to that vertex).

Reply:
It's now: `When a new kmer \textit{or a non-unique kmer (even if it is already on the graph)} appears, a new vertex is added to the graph; otherwise an edge is added (or the multiplicity of an existing edge incremented) from the preceding kmer to the current one.`

Reply:
Is a bit more complex than that I'm afraid.... and old non-unique kmer may or may not result in a new vertex.
When a new vertex appear this is added to the graph, when and old vertex appear a few things may happen. If it is unique then no new vertex is added (and a edge is draw from the current vertex to this (old) next), if it is non-unique then if there is already a edge between the current vertex and the old next then we only increase the weight on that edge otherwise a new vertex and edge are added to the graph.